### ISSUE TITLE
Scala time-out

### ISSUE THREAD

**Issue by g964** [Codewars/codewars.com#1289]
CW Scala is so slow that 27 second is not that much...

**Comment by kazk** [comment]
I agree it's slow and it's frustrating.
I can't do much about it at the moment, but here's some reasons and what we're working on:

#### Underpowered machines
Machines used for Codewars isn't really powerful and the resources are limited due to the current backend design. I think this used to work fine when Codewars only supported few languages, but it's been suffering as the number of supported languages increased, especially those demanding more resources.

We're building a new system and I'm hoping this will allow us to reduce the number of machines required and use less machines with more capacity while keeping the cost at similar level.

#### Using Gradle is inefficient
I used it to implement the runner because it was used for Java already and it was easiest to ship Groovy/Kotlin/Scala runner. But it does much more than we need.

On the new system we've been working on, I've added a Scala runner without Gradle and it reduces the time to almost half. In this method, managing dependencies is harder and it required more time to implement so it's a trade-off.

#### Compiling Scala the first time is slow
As far as I know, the compiler itself is slow when compiling the code the first time. This is understandable because usually this happens only once and compiling is done incrementally thereafter. On Codewars, this is a huge cost because we do this every time the code is submitted.

I don't think there's anything we can do about this.

#### Beta language
Groovy/Kotlin/Scala is still considered beta and not much time has been put into them.

**Comment by g964** [comment]
Lots of thanks for the explanations but isn't it possible to increase the 27 s until the new system is in place?
I tried an easy new Scala kata with only 2 fixed tests:
I get : "Time: 22367ms" and "Completed in 1190ms". Why this big difference between "Time" and "Completed"?

**Comment by kazk** [comment]
> isn't it possible to increase the 27 s until the new system is in place?

Maybe.
```javascript
    scala: 27000,
```
https://github.com/Codewars/codewars-runner-cli/blob/dbf943d62c8d6ed451d27ac1817c0f13b1de558c/lib/config.js#L21

You can change this line and make a PR to propose. Jake has been busy with other tasks so I don't know how soon it'll get deployed though.

---

> I get : "Time: 22367ms" and "Completed in 1190ms". Why this big difference between "Time" and "Completed"?

"Completed in" is based on the time reported by the test framework and it's the time it took to **run** the test.

"Time" is how long it took in total and measured by the runner. There're some inconsistencies between languages, but for Scala, this includes the time it took to compile. The maximum time allowed is based on this.

**Comment by kazk** [comment]
As I wrote in https://github.com/Codewars/codewars.com/issues/1336#issuecomment-378338319 I'm going to switch to using ScalaTest directly soon. This should allow most kata to finish under 8 seconds.

**Comment by kazk** [comment]
We've switched to using ScalaTest. I'll be updating the existing content once the change is applied to the frontend.

**Comment by Voileexperiments** [comment]
@kazk I'm now trying to get my fix of https://www.codewars.com/kata/bagels published, and it always times out *and fails*. It seems that the `https://www.codewars.com/api/v1/deferred/:dmid` endpoint just keeps returning `{"_type":"progress","value":"queued"}` for a minute until then kills the republish and returns `{"success":false,"reason":"not found"}`. Is there something in the queue blocking the republish?

**Comment by kazk** [comment]
Not sure. I couldn't find anything wrong. But I don't know much about Codewars itself (only Jake knows). I'm hoping that I can start reading the codebase soon and help fixing.

We do know that Codewars has **lots** of issues (as you've been reporting :)) and care, but we simply don't have enough resources (people) at the moment.
